# Proyecto 3 - Limpieza de datos




*   DIEGO FERNANDO VILLAIZAN MELO
*   MARIA FERNANDA MORA GUERRA
*   JESUS ANTONIO REYES ACOSTA

Realice una limpieza a la tabla dataset1. En un notebook lleve a cabo los siguientes ítems:


In [ ]:
import pandas as pd 
import numpy as np

**Carga de datos:**

In [ ]:
df = pd.read_csv ("https://raw.githubusercontent.com/labeconometria/MLxE/main/proyectos1er/dataset_1.csv")

**Exploracion inicial**

In [ ]:
df.head()
#primeros 5 datos de la tabla

user_id  gender      wage  purchased   birth_day
0      217  Female       NaN          0  1983-08-06
1      226  Female  146500.0          1  1972-01-28
2      697    Male       NaN          1  1981-12-12
3      743  Female       NaN          0  1983-05-26
4       63  Female  140500.0          1  1967-11-02

In [ ]:
df.tail() 
#Ultimos 5 ultimos datos

user_id  gender      wage  purchased   birth_day
1169      636    Male   76500.0          0  1970-08-26
1170      320  Female  143500.0          1  1970-04-07
1171      136  Female       NaN          1  1968-12-28
1172      721    Male  148000.0          1  1993-11-23
1173      471     man   60500.0          0  1984-06-02

In [ ]:
df.info() 
#datos de las columnas


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1174 entries, 0 to 1173
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   user_id    1174 non-null   int64  
 1   gender     1174 non-null   object 
 2   wage       930 non-null    float64
 3   purchased  1174 non-null   int64  
 4   birth_day  1174 non-null   object 
dtypes: float64(1), int64(2), object(2)
memory usage: 46.0+ KB


El comando .info() nos entrega en detalle los datos que registra cada observacion, el total de columnas y lo que indica cada una de estas.

**Revisión de duplicados explícitos e implícitos:**

In [ ]:
print(f'Numero de datos totales con duplicados : {len(df)}')
df = df.drop_duplicates()  # Se eliminan duplicados explicitos
print(f'Numero de datos sin duplicados : {len(df)}')

Numero de datos totales con duplicados : 1174
Numero de datos sin duplicados : 1000


In [ ]:
for columna in df.columns:    
  print(df[columna].unique())
# se debe hacer un analisis detallado para ver si hay duplicados implicitos


[ 217  226  697  743   63   92  971  777  753  211  891  817  587  889
  126  238  733  483  463  139  994  473  316  662  578  180  631  129
  688  341  382   54  998  961  583  128  542  654  950  367  243  488
   80  307  793  269  607  319  702  903  745    1  679  757  689  440
  568  922  701   64  346  775  432  551  107  674  283  321  653  393
  962  357   73  771  304  429  703   69  363  308  544  419  599  917
  744  555  867   82  120  175  279  329  567  188  833  579  859  142
  791  589  276  704   19  729  815  651  850  335  200  944  947  621
  204  780  902  417  420  462  750  273  522  907  360  968  672  262
  324  881  285  371  554  720  250  151  914  625  862  106  582  192
  766  687  983  841  242  479  228  478  178  206  616  931  286  788
  378   23  369  428  513  541  433  476   90  127  834  553  386  915
  822   25  540  957  987  213  591   55  742  698  425  576  893  796
  974  160  100  581  716  300  695   27  911  942  933  533  690  920
  149 

In [ ]:
df['gender'] = df['gender'].replace({'woman':'Female','man':'Male'})
df['gender'].unique()
#Se modifico la variable "gender"

array(['Female', 'Male'], dtype=object)

Los datos duplicados los encontramos como implicitos y explicitos, en el cual fue el primer paso de la limpieza los cuales eran datos explicitos, dejando como resdultado un total de 1000 observaciones.
Para visualizar los duplicados implicitos es necesario verificar si los valores tienen valores duplicados, en este caso se corrige con el comando replace een la variable "gender".

# Transformacion de los datos

In [ ]:
df.dtypes
#Observamos los datos 

user_id        int64
gender        object
wage         float64
purchased      int64
birth_day     object
dtype: object

In [ ]:
df['birth_day'] = pd.to_datetime(df['birth_day']) 
df['purchased'] = df['purchased'].astype('category')
# Modificamos el tipo de dato en la variable birth_day

# Generando variable Age

In [ ]:
df['age'] = [2022 - año for año in df['birth_day'].dt.year]
df['age'].describe()
# A partir de la variable de fecha de nacimiento, podremos calcular la edad de los individuos la cual podfremos usar de una mejor manera para realizar las categorizaciones por rango de edad.

count    1000.000000
mean       40.106000
std        10.707073
min        18.000000
25%        32.000000
50%        40.000000
75%        48.000000
max        63.000000
Name: age, dtype: float64

# **Missing values**

In [ ]:
n = len(df)
print('Porcentaje de Missing values por variable')
print(' ')
for columna in df.columns:
  porcentaje = df[columna].isna().sum()/n
  print(f'{columna}: {porcentaje*100} %')

Porcentaje de Missing values por variable
 
user_id: 0.0 %
gender: 0.0 %
wage: 21.4 %
purchased: 0.0 %
birth_day: 0.0 %
age: 0.0 %


In [ ]:
df['wage'].describe()


count       786.000000
mean      72580.152672
std       34268.329521
min       15000.000000
25%       46125.000000
50%       71750.000000
75%       90000.000000
max      152500.000000
Name: wage, dtype: float64

In [ ]:
df['wage'] = df['wage'].fillna(df['wage'].mean())

In [ ]:
df['wage'].describe()

count      1000.000000
mean      72580.152672
std       30377.009614
min       15000.000000
25%       53500.000000
50%       72580.152672
75%       83000.000000
max      152500.000000
Name: wage, dtype: float64

La variable wage se encontro con el 21.4 % de datos faltantes con el comando .isna(), este porcentaje tan elevado no lo podemos nos afecta los resultados observacionales, lo cual se decide reeemplzar los valores faltantes por lla media de la misma columna.

# **outliers**

In [ ]:
rango_iq = df['wage'].quantile(0.75)  - df['wage'].quantile(0.25)

limite_superior = df['wage'].quantile(0.75) + (1.5 * rango_iq)
limite_inferior = df['wage'].quantile(0.25) - (1.5 * rango_iq)

print('Analizis outliers Variable Wage')
print(' ')
print(f'Limite Superior = {limite_superior}')
print(f'Limite Inferio = {limite_inferior}') 

Analizis outliers Variable Wage
 
Limite Superior = 127250.0
Limite Inferio = 9250.0


In [ ]:
copia = df.copy()

In [ ]:
n_antes = len(copia)
df = df[(df['wage']<=limite_superior) & (df['wage']>=limite_inferior)]
n_despues = len(df)
print(f'Se eliminaron {n_antes - n_despues} datos ({((n_antes-n_despues)/n_antes) * 100}% del total)')

Se eliminaron 77 datos (7.7% del total)


En este caso ajustamos los datos atipicos que se encuentren fuera del rango interquantilico, estos datos pueden afectar significativamente el analisis.

# **Manejo de memoria**

In [ ]:
df.memory_usage()

Index        7384
user_id      7384
gender       7384
wage         7384
purchased    1047
birth_day    7384
age          7384
dtype: int64

In [ ]:
lista = ['purchased','age']
for columna in lista:
  df[columna] = df[columna].astype('int8')

df['gender'] = df['gender'].astype('category')

In [ ]:
df.memory_usage()

Index        7384
user_id      7384
gender       1047
wage         7384
purchased     923
birth_day    7384
age           923
dtype: int64

debemos reducir la memoria que ocupa la tabla, los numeros enteros pasan a formato "int-8", el cual ocupa menos memoria .

# **Clasificacion de datos**
(
  grupos etarios)

In [ ]:
df['age'].describe()

count    923.000000
mean      39.631636
std       10.728293
min       18.000000
25%       32.000000
50%       39.000000
75%       47.000000
max       63.000000
Name: age, dtype: float64

In [ ]:
df['grupo_etario'] = df['age']
df['grupo_etario'] = np.where((df['grupo_etario']>=14) & (df['grupo_etario']<=26) , -111 , df['grupo_etario']) # jovenes
df['grupo_etario'] = np.where((df['grupo_etario']>26) & (df['grupo_etario']<=59) , -222 , df['grupo_etario']) # Adultos 
df['grupo_etario'] = np.where(df['grupo_etario']>59 , -333 , df['grupo_etario']) # Tercera Edad
df['grupo_etario'] = df['grupo_etario'].replace({-111:'Joven',-222:'Adulto',-333:'Tercera Edad'})

In [ ]:
df['grupo_etario'].value_counts()

Adulto          768
Joven           114
Tercera Edad     41
Name: grupo_etario, dtype: int64

# **grupo de ingreso**

In [ ]:
df['wage'].describe()

count       923.000000
mean      66897.781876
std       24003.697750
min       15000.000000
25%       51500.000000
50%       72580.152672
75%       79000.000000
max      126500.000000
Name: wage, dtype: float64

In [ ]:
df['grupo_ingreso'] = df['wage']
df['grupo_ingreso'] = np.where(df['grupo_ingreso']<=42000,-111,df['grupo_ingreso'])
df['grupo_ingreso'] = np.where((df['grupo_ingreso']>42000) & (df['grupo_ingreso']<=84000),-222,df['grupo_ingreso'])
df['grupo_ingreso'] = np.where(df['grupo_ingreso']>84000,-333,df['grupo_ingreso'])

df['grupo_ingreso'] = df['grupo_ingreso'].replace({-111:'Ingreso Bajo',-222:'Ingreso Medio',-333:'Ingreso Alto'})
df['grupo_ingreso'].value_counts()

Ingreso Medio    593
Ingreso Bajo     165
Ingreso Alto     165
Name: grupo_ingreso, dtype: int64

Para esta parte generamos dos varialbes nuevas en la tabla, las dos buscan categorizar la poblacion en determinados grupos segun las condiciones establecidas.


*   Grupo etario, la cual segmenta por edades joven (14-26 años), adulto, (27-59años), tercera edad (60+ años)
*   Ingresos del individuo, ingreso bajo (<=42000), ingreso medio (42000-84000), ingreso alto (+84000) 



# Tablas Pivot:

In [ ]:
df['conteo'] = 1
tabla1 = pd.pivot_table(df,index='grupo_etario',columns='purchased',values='conteo',aggfunc='sum').fillna(0)
tabla1['cantidad'] = df['grupo_etario'].value_counts().values
tabla1['Porcentaje de Compras por grupo'] = tabla1[1]/tabla1['cantidad'] *100
# tabla1[['Porcentaje de Compras por grupo']]
tabla1

purchased         0      1  cantidad  Porcentaje de Compras por grupo
grupo_etario                                                         
Adulto        473.0  295.0       768                        38.411458
Joven         114.0    0.0       114                         0.000000
Tercera Edad    3.0   38.0        41                        92.682927

Esta tabla nos muestra el porcentaje de compradores dentro de los rangos de edad, para la tercera nos indica quwe el 92% de las personas de la tercera edad son compradores; sin embargo no es tan significativo ya que tan solo tenemos 41 individuos.

In [ ]:
tabla2 = pd.pivot_table(df,index='grupo_ingreso',columns='purchased',values='conteo',aggfunc='sum')
tabla2['cantidad'] = [165,165,593]
tabla2['Porcentaje de Compras por grupo'] = tabla2[1]/tabla2['cantidad'] *100
# tabla2[['Porcentaje de Compras por grupo']]
tabla2

purchased        0    1  cantidad  Porcentaje de Compras por grupo
grupo_ingreso                                                     
Ingreso Alto    52  113       165                        68.484848
Ingreso Bajo    82   83       165                        50.303030
Ingreso Medio  456  137       593                        23.102867

Para este caso podemos detallar el nivel de compradores de acuerdo al ingreso, es interesante ver el resultado de los ingreso altos en los cuales el 68% son compradores a comparacion del ingreso medio, que tiene el 23% pero la cantidad de individuos es mayor.

In [ ]:
tabla3 = pd.pivot_table(df,index='gender',columns='purchased',values='conteo',aggfunc='sum')
tabla3['cantidad'] = tabla3[0] + tabla3[1] 
tabla3['Porcentaje de Compras por grupo'] = tabla3[1]/tabla3['cantidad'] *100
# tabla3[['Porcentaje de Compras por grupo']]
tabla3

purchased    0    1  cantidad  Porcentaje de Compras por grupo
gender                                                        
Female     295  180       475                        37.894737
Male       295  153       448                        34.151786

Podemos ver que las mujeres tienen un porcentaje mayor de compras a los hombres.

In [ ]:
tabla4 = pd.pivot_table(df,index=['grupo_etario','grupo_ingreso','gender'],columns='purchased',values='conteo',aggfunc='sum')
tabla4['cantidad'] = tabla4[0] + tabla4[1]
tabla4['Porcentaje de Compras por grupo'] = tabla4[1]/tabla4['cantidad'] * 100
# tabla4[['Porcentaje de Compras por grupo']]
tabla4

purchased                            0   1  cantidad  \
grupo_etario grupo_ingreso gender                      
Adulto       Ingreso Alto  Female   25  55        80   
                           Male     15  48        63   
             Ingreso Bajo  Female   27  41        68   
                           Male     26  33        59   
             Ingreso Medio Female  187  64       251   
                           Male    193  54       247   
Joven        Ingreso Alto  Female    3   0         3   
                           Male      8   0         8   
             Ingreso Bajo  Female   15   0        15   
                           Male     14   0        14   
             Ingreso Medio Female   35   0        35   
                           Male     39   0        39   
Tercera Edad Ingreso Alto  Female    1   6         7   
                           Male      0   4         4   
             Ingreso Bajo  Female    0   5         5   
                           Male      0   4         4   
             Ingreso Medio Female    2   9        11   
                           Male      0  10        10   

purchased                          Porcentaje de Compras por grupo  
grupo_etario grupo_ingreso gender                                   
Adulto       Ingreso Alto  Female                        68.750000  
                           Male                          76.190476  
             Ingreso Bajo  Female                        60.294118  
                           Male                          55.932203  
             Ingreso Medio Female                        25.498008  
                           Male                          21.862348  
Joven        Ingreso Alto  Female                         0.000000  
                           Male                           0.000000  
             Ingreso Bajo  Female                         0.000000  
                           Male                           0.000000  
             Ingreso Medio Female                         0.000000  
                           Male                           0.000000  
Tercera Edad Ingreso Alto  Female                        85.714286  
                           Male                         100.000000  
             Ingreso Bajo  Female                       100.000000  
                           Male                         100.000000  
             Ingreso Medio Female                        81.818182  
                           Male                         100.000000

En este caso tenemos una tabla pivot que nos muestra el resultado por edad, ingresos y genero.
Podemos concluir que el grupo mas comprador es el de mujeres adultas de ingreso medio.

In [ ]:
tabla5 = pd.pivot_table(df,index='grupo_etario',values='wage',aggfunc=np.mean)
tabla5 = pd.concat([tabla5,pd.pivot_table(df,index='grupo_etario',values='wage',aggfunc=np.median)['wage']],axis=1)
tabla5.columns = ['Promedio Ingreso','Mediana Ingreso']
tabla5

Promedio Ingreso  Mediana Ingreso
grupo_etario                                   
Adulto            67890.008747     72580.152672
Joven             60221.441007     72580.152672
Tercera Edad      66875.162912     72580.152672

In [ ]:
tabla6 = pd.pivot_table(df,index='gender',values='wage',aggfunc=np.mean)
tabla6 = pd.concat([tabla6,pd.pivot_table(df,index='gender',values='wage',aggfunc=np.median)['wage']],axis=1)
tabla6.columns = ['Promedio Ingreso','Mediana Ingreso']
tabla6

Promedio Ingreso  Mediana Ingreso
gender                                   
Female      67224.162314     72580.152672
Male        66551.731189     72580.152672

estas tablas nos muestras la mediana del ingreso por genero y grupo etarios, podemos decir que los los hombres perciben menos ingresos que las mujeres.

# **Conclusion**

La exploracion inicial y limpieza de los datos, es un primer proceso importante para el analisis ya que dentro de estos datos nos encontramos con datos faltantes, repetidos que pueden cambiar significativamente los resultados a los que queremos llegar.

La segmentacion que se le da a la informacion para llegar a analisis es muy interesante pues podemos tener varios analisis de una poblacion, tales como los hicimos en el ejercicio en los cuales comparamos los ingresos por edades o genero, el cual puede determinar de una manera precisa cierta informacion de determinado lugar o empresa. 